File: DeploymentTargetTest/BrightnessController.m
// BrightnessController.m

#import "BrightnessController.h"
#import <IOKit/IOKitLib.h>
#import <IOKit/graphics/IOGraphicsLib.h>

//#import <IOKit/i2c/IODALAccess.h>

@implementation BrightnessController

/**
 @param brightnessValue this value controls the screen brightness with 0 being minimum brightness and 1 being maximum brightness
 */
+ (void)setScreenBrightness:(float)brightnessValue {
    io_iterator_t iterator;
    kern_return_t result = IOServiceGetMatchingServices(kIOMasterPortDefault, IOServiceMatching("IODisplayConnect"), &iterator);
    
    // If we were successful
    if (result == kIOReturnSuccess)
    {
        io_object_t service;
        while ((service = IOIteratorNext(iterator))) {
            // Print the current brightness of the screen
            float current_brightness;
            IODisplayGetFloatParameter(service, kNilOptions, CFSTR(kIODisplayBrightnessKey), &current_brightness);
            printf("current brightness: %.0f%%\n", current_brightness * 100);

            IODisplaySetFloatParameter(service, kNilOptions, CFSTR(kIODisplayBrightnessKey), brightnessValue);
            
            // Print the custom brightness of the screen
            IODisplayGetFloatParameter(service, kNilOptions, CFSTR(kIODisplayBrightnessKey),
                                                &current_brightness);
            printf("new custom brightness: %.0f%%\n", current_brightness * 100);

            // Let the object go
            IOObjectRelease(service);
        }
    }
}

@end

File: DeploymentTargetTest/BrightnessController.swift
//
//  BrightnessController.swift
//  DeploymentTargetTest
//
//  Created by Denis on 01.08.2026.
//

import Foundation
import IOKit

/// Utility that lets you set the screen brightness via the IOKit “IODisplay” interface.
class BrightnessController {

    /// Sets the screen backlight level.
    ///
    /// - Parameter brightnessValue: Normalized value between 0 (off) and 1 (maximum).
    static func setScreenBrightness(_ brightnessValue: Float) {
        var iterator = io_iterator_t()

        // Find all services that expose a display connection
        let result = IOServiceGetMatchingServices(
            kIOMasterPortDefault,
            IOServiceMatching("IODisplayConnect"),
            &iterator
        )

        guard result == KERN_SUCCESS else { return }

        while case let service = IOIteratorNext(iterator), service != 0 {
            var currentBrightness: Float = 0.0

            // Read the current brightness
            IODisplayGetFloatParameter(
                service,
                0,
                kIODisplayBrightnessKey as CFString,
                &currentBrightness
            )
            print("current brightness: \(Int(currentBrightness * 100))%")

            // Apply the new value
            IODisplaySetFloatParameter(
                service,
                0,
                kIODisplayBrightnessKey as CFString,
                brightnessValue
            )

            // Verify the change
            IODisplayGetFloatParameter(
                service,
                0,
                kIODisplayBrightnessKey as CFString,
                &currentBrightness
            )
            print("new custom brightness: \(Int(currentBrightness * 100))%")

            IOObjectRelease(service)
        }
    }
}
